[TEP-0076] Propose array results and indexing 🐠

bobcatfish opened the pull request:
This TEP proposes adding more support for array params by adding
array results as well as the ability to index into arrays.

It refers to TEP-0075 which will be added in a separate commit, which
proposes adding support for dictionary types.

Related issues:
* [pipelines#1393 Consider removing type from params (or _really_ support types)](https://github.com/tektoncd/pipeline/issues/1393)
* [pipelines#3255 Arguments of type array cannot be access via index](https://github.com/tektoncd/pipeline/issues/3255)

/kind tep

bobcatfish commented:
p.s. why 76 you might ask? what happened to 74 and 75 XD? well you see, i started working on 74, then i realized i needed 75 (dictionary params + results) and THEN i realized i needed 76, so here we are 😆  (should see 74 + 75 as well asap!!)

bobcatfish commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+    - environments:
+        - 'staging'
+        - 'qa'
+        - 'prod'
```
whoops thats not right, it would look like:

```yaml
  params:
    - name: environments
       value:
        - 'staging'
        - 'qa'
        - 'prod'
```

sbwsg replied:
I think it would require `type: array` as well? At least for task params this is required I think.

bobcatfish replied:
whoops! thanks :D

bobcatfish replied:
i just realized i dont think i included a single example of how you actually specify a result with that type either XD

bobcatfish replied:
... and then i forgot to fix it again XD

bobcatfish replied:
aaaaand the values should be in the pipelinerun not the pipeline .. wheeee.....

bobcatfish replied:
i think i finally fixed it XD

chhsia0 commented:
/assign @sbwsg @pierretasci

pierretasci commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+  param of another Task in the pipeline which has an array param
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
```
I don't think the benefit of the use case provided is addressed by these requirements. The key missing piece for me is dynamism. Already knowing the size of your array upfront is great and all but I think it is worth exploring with this effort how we can support variable sized arrays and perhaps even task generation based on some sort of template.

afrittoli replied:
Dynamism here is supported using task loops, since you don't need to know the size of the array there, you just pass the whole array and the custom controller does the rest.
Eventually we should get loops into the main API to support what you're asking for without the need for custom tasks.

I'm not sure tbh what the use case for indexing is?

afrittoli replied:
After the discussion on https://github.com/tektoncd/pipeline/issues/4097, I think I see a good use case for indexing - i.e. especially for results, there are cases where it makes sense of processing an array of results instead of feeding it to a loop of tasks directly.

This will require being able to know the size of the array though, so it feels like indexing and length should go together?

bobcatfish replied:
I'm assuming this is addressed via the conversation at https://github.com/tektoncd/community/pull/477#discussion_r673508137 but plz re-open if not!

pierretasci commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+This adds some additional complexity in that there are two paths to which you could write results (the string file
+and the json file), but this option has these advantages:
+
+* Completely backward compatible with current use of the results file
```
Can we not just use the same path and under-the-hood always use json with a `format` key telling us if the content is meant to be read out as a string, dict, or array?

bobcatfish replied:
that is a GREAT idea! :D

pierretasci commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+
+* What if the file contains json we don't support? (e.g. we don't yet support dictionaries, and wouldn't
+  initially support arrays of arrays)
+    * The TaskRun (or Pipeline Task) would fail
```
We could at least partially validate this on the creation of the resource too

bobcatfish replied:
could you explain a bit more? I'm not sure how we'd know until runtime that something was written to the file in an unexpected format

pierretasci replied:
Yeah that is what I mean. We can validate it at runtime and potentially fail the pipeline

bobcatfish replied:
I'm adding a section "Validating JSON results at runtime" to cover this in more detail and moving this list out of the notes/caveats section

pierretasci commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+        - 'qa'
+        - 'prod'
+...
+  tasks:
+    - name: deploy
+      params:
+        - name: environment
+          value: '$(params.environments[0])'
```
this is where I think dynamism could be super cool. What if we did:
```suggestion
  tasks:
    - name: deploy
      params:
        - name: environment
          value: '$(params.environments[])'
```

Having the non-indexed array there, similar to `jq` would produce a copy of task deploy for each valid in the input `params.environments`. Seems fairly straight-forward and a reasonable addition to the api.

vdemeester replied:
I disagree there, or at least in the scope of the TEP. This might be something that appear confusing to the users, and most importantly, I would like to make sure we really need this (and custom task are not enough) before commiting to that level of complexity.

pierretasci replied:
I can think of many use cases related to test sharding where this would be super useful. We also have many "build these X services and deploy them to an ephemeral cluster" uses cases. All involve a ton of copy+pasta of the same exact spec with different params

vdemeester replied:
Right, but as of today, this is what `task-loop` custom task is there for, isn't it ?

bobcatfish replied:
@pierretasci  100% agree with the test sharding use case and i have a languishing todo list item about trying to write up a TEP about better supporting test sharding specifically (which ultimately would lead to needing looping support)

I also agree with @vdemeester though that I don't personally see that as being exaclty within the scope of this particular TEP to add looping support BUT I _do_ see array result support as a requirement for being able to support looping (and sharding) - and maybe that's what you're getting at @pierretasci (not that this TEP includes implementing looping, but that we need this TEP for looping and sharding is a great use case to list to justify adding array results)

pierretasci replied:
Well I was specifically advocating for supporting it directly in the Tekton API because I think that unlocks many more use-cases but I can also see the argument against.

bobcatfish replied:
i am 💯  in favor of looping in the pipelines api at this point but i think we need to explore the use cases and alternatives a bit (and hopefully @vdemeester will find them convincing :D )

bobcatfish replied:
https://github.com/tektoncd/community/pull/600 🎉 🎉 🎉

pierretasci reviewed (commented):
Overall, I think this is a more than worthwhile change and use case. My only comment is that I think we should take it one step further but otherwise, this is LGTM from me.

/LGTM

vdemeester commented:
/hold

sbwsg reviewed (commented):
I think array params require the an author to explicitly declare the type:

```yaml
spec:
  params:
  - name: foo
    type: array
```

If type isn't specified then the default is assumed to be `string`. A taskrun cannot supply an array of results when task doesn't specify param `type`. Validation fails with something like `invalid input params for task mytask: param types don''t match the user-specified type: [foo]`.

From my pov it would make sense that we would also require `type` when a task declares results:

```yaml
kind: Task
spec:
  results:
  - name: foo
    type: array
```

And eventually `type: dictionary` or similar.

In which case, I'd just echo what @pierretasci already suggested - we might not need a different result path. The result's type in the task definition would explicitly tell the entrypoint/controller how the file content should have been written, should be parsed, and whether it's valid or not. I think this would be totally backwards compatible since `type` would default to `string`.

wdyt? I might be missing something here.

sbwsg commented:
Ah ok, sorry, just read the `dictionary` proposal and I see now that we're [discussing `type` being deprecated in favor of `schema`](https://github.com/tektoncd/community/pull/479/files#diff-b7a822637583edb5e250858789d03ae285ff09dc2c63354fc4f995c588afc581R268-R274).

sbwsg commented:
My questions are entirely focused around the implementation and since we're only at the proposal stage I'd def like to see this move ahead!

/approve

vdemeester commented:
/assign

afrittoli commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+   splits a string). For example, in tektoncd/plumbing we have a directory of
+   [images](https://github.com/tektoncd/plumbing/tree/main/tekton/images). Each directory contains a Dockerfile -
+   in a Pipeline that builds all images in this repo, we would want to: 1) determine which Dockerfiles need to be built
+   and 2) for each of those Dockerfiles, build the image. Looping support would allow us to use a Task which builds
```
I would extend this by saying that we also need to be capture results from a loop.
It we build an array of container images, we need to be able to get back an arrays of results.
In general, it is not possible to use a task with results in a loop because of this today.

bobcatfish replied:
Ahh interesting, this makes me think of 2 more use cases we could add (let me know if this is what you're thinking too or if I'm missing the point):

1. If the "loop" is inside of the Task, e.g. if a task knew how to take an array of dockerfile locations and build each of them, then I think I agree that you'd need to be able to produce an array of results in order to contain the results of each iteration (maybe i can add this as a use case and use our [release publish task](https://github.com/tektoncd/pipeline/blob/main/tekton/publish.yaml) as an example - it builds mulitple images AND it loops over them and publishes them to multiple locations resulting in multiple URLs)
2. It seems like it's important to let custom tasks (via Runs) emit array results because you don't know what kind of looping or other logic they might contain - e.g. [the existing looping custom task mentions the fact that it doesnt create results as a limitation](https://github.com/tektoncd/experimental/tree/main/task-loops#limitations) - and with the current interface, it would have to combine the results of each task into a string

(the last question would be what does the results interface look like if we support looping directly in the Tekton API but we can leave that for a looping focused TEP :D)

bobcatfish replied:
or maybe a better way to state it, and maybe this is more what you were getting at @afrittoli, is supporting looping where you both need to **consume** array results (to loop over) and **produce** array results (from those loops)

afrittoli commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+
+### Use Cases
+
+1. Looping (i.e. providing values for an interface such as the task-loop custom task
```
Since looping is not part of the core API today, I wonder if we could add another use case here.
Even if an array is not used as an input to a loop, it might be used as args for another task.
I could have a task that builds the args for the next task, by producing an array of parameters.

vdemeester replied:
+:100: on what @afrittoli just wrote. Given the current feature set of pipeline, I think that should appear as the main use case 👼🏼

afrittoli commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+
+* Add two missing pieces of array support
+  * As results
+  * Indexing into arrays
```
I think there's no use case defined to support the indexing goal.
If we add indexing, we might need to introduce an operator that allows checking the size of an array too?

vdemeester replied:
Good point 😅

bobcatfish replied:
another +1 to including the size as a variable replacement, that could be a way to solve https://github.com/tektoncd/pipeline/issues/4097

(will need to check if json schema provides something like this or if we would need to pave new ground)

afrittoli replied:
@bobcatfish This seems like a good use case to have in the TEP! 👍

bobcatfish replied:
circling back to this, I forgot that @sbwsg and I had already discussed this a bit and I'm back to leaving this out of scope for now (discussion: https://github.com/tektoncd/community/pull/477#issuecomment-901308105 TL;DR not clear how to support this for other types e.g. objects, or where we'd draw the line - also we can always add this later!) - it's currently listed in the alternatives section of the doc

afrittoli commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+     would mean not allowing the ability to get individual values out of those dictionaries which would significantly
+     limit the number of Tasks this would be compatible with (i.e. we should be consistent between arrays and
+     dictionaries)
+   * Con: Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that
+     only take string parameters and this limits the interoperabiltiy between Tasks.
```
It makes sense, I kind of struggle to imaging an actual example though.
I would need an array parameter where certain positions in the array hold special meaning I guess?

afrittoli reviewed (commented):
Thank you! Just added a few comments, mostly I'm not 100% sure about indexing, but in general this looks good!

jerop commented:
/assign

vdemeester commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+
+## Motivation
+
+* See [use cases](#use-cases) - use cases for looping support in pipelines often need an array input;
+  being able to produce an array from a Task to feed into a loop would enable some interesting CD use cases
```
I'm still not sold on having loop support in pipeline itself (aka as a custom task looks just fine to me).

pierretasci replied:
Obviously, it could be a custom tasks but I think this is such a powerful feature that it deserves to be in the core API. If for no other reason than once you add the dynamism of looping, you add an extra dimension which Tekton can be extended on. Once you have looping, you can now support trivial replication fan-out, or conditional looping with when expressions, etc all with existing primitives.

afrittoli replied:
Having native loop support in the core API does not mean that we cannot have a dedicated controller/CRD for it still.
Loops are probably a very good use case for TEP-0044, where we might not want to have one dedicated Pod for each iteration. A `TaskGroup` controller could decide what is the best `TaskRuns` <-> `Pod` ratio based on the inputs.

bobcatfish replied:
we've got a TEP for this now! :D 🎉 https://github.com/tektoncd/community/pull/600

vdemeester commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+
+## Proposal
+
+Currently we treat the contents of
+[a Task's result file(s)](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+as a string. The file is located at `$(results.resultName.path)` which expands to `/tekton/results/resultName`.
+In this proposal, we add the option to write to a json file instead, available at `$(results.resultName.jsonPath)`
+which expands to `/tekton/results/resultName.json`. The contents of this file would be treated as json and in addition
+to supporting a string value in this file, we sould support json arrays.
```
Any reason for going `json` vs another format ?

vdemeester replied:
And potential issues ? (if someone write a complex json file, what do we do ?)

bobcatfish replied:
@vdemeester do you have any other formats you'd suggest? off the top of my head all I can think of are:

* yaml (i dont think there is a syntax like jsonpath available for yaml? i.e. it'd be harder to expose via variable replacement - or we'd need to convert from yaml to json internally)
* protobuf (this would dramatically increase the difficulty in emitting structured results from tasks - either we'd see every Task including a step with protobuf tools in it or we'd need to include some by default)
* xml (seems more verbose than we'd need?)
* csv (would also be easy to use but afaik it wouldn't support more complex types like dicts? or we'd need to add some kind of syntax on top of the csv format to define more complex types)

as for why json I'd say because:
* it's human readable and so it's easy for folks to write json content without needing extra tools
* we've added json specific support to other parts of our APIs (e.g. [in triggers](https://github.com/tektoncd/triggers/blob/f2380d7d33666547ee7ed965730cf7dbc1aa5421/docs/triggerbindings.md#accessing-data-in-http-json-payloads))
* yaml and json are the two formats that k8s is using so I think it makes sense to use one of those

bobcatfish replied:
> if someone write a complex json file, what do we do 

i think we'd simply bail out of evaluating the json as soon as it doesn't match the expected schema (which i think the "notes/caveats" tries to address but ill try to make it more clear) and also given the size limits, there's a limit on how complex the json file can be (maybe you have some specific scenario in mind tho?)

bobcatfish replied:
| if someone write a complex json file, what do we do

ill add something to the testing section as well to make sure this is covered (a deeply nested dictionary is the most complex scenario i can think of?)

vdemeester commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+
+## Drawbacks
+
+* Increases the size of TaskRuns (by including arrays in termination messages and results)
```
Any possible problem if the arrays is huge (a "from a task" attach that would write a 10M array json file for example) ?

bobcatfish replied:
if [the termination message ended up being > 4096 bytes it would get truncated](https://kubernetes.io/docs/tasks/debug-application-cluster/determine-reason-pod-failure/#customizing-the-termination-message) (and end up being invalid json and the taskrun would fail)  (unless as suggested in https://github.com/tektoncd/pipeline/issues/4012 we adopt an alternative) - anyway ill mention this in the TEP

vdemeester commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,321 @@
+1. Instead of adding support for a `.json` file in addition to the existing path, we use the same file and parse it as
+  json.
+    * Pro: Avoids the complication of two different paths for result files
+    * Con: A backwards incompatible change for Tasks which are currently writing json to the results file and treating
+      it as a string. The feature would be behind an alpha flag so they wouldn't be impacted immediately but eventually
+      they would need to update their Tasks to escape the json.
```
This is definitely a valid concern, not really highlighted above I think. What happens to those use-cases once we rely solely on the json syntax for results ? How a user can pass a json as a string result ?

bobcatfish replied:
the original proposal was that users when users wanted to emit something other than a string, they'd write it to the `.json` file - otherwise anything they write to the results file would be treated as just a string so they wouldn't need to change anything (i.e. we'd never rely soley on json syntax for results - we'd support both the current "string" syntax and the json syntax)

@pierretasci and @sbwsg pointed out an improvement we could make where we can (have our 🍰  and eat it to) support both without needing to introduce a new results file - we'd rely on how the Task defines the result in its spec to tell us if what we encounter in the file should be treated as json or treated as a string)

bobcatfish commented:
Multiple ppl ( @pierretasci and @afrittoli I think at least) questioned the use case for indexing and I've been racking my brain to come up with a use case and there's pretty much only one (pattern) I can think of. It's not super compelling though (the looping use cases are way more compelling), so I'd be okay with removing the indexing part part of this proposal and limiting it to just results if others feel strongly.

that being said I still think supporting indexing makes the package of array support complete - and if we do add [dict/object support](https://github.com/tektoncd/community/pull/479) it seems lopsided to me to support something like `$tasks.foo.results.mydict["bar"]` but not `$tasks.foo.results.myarray[1]

I'll add this in more detail under use cases, but the only use case I can come up with for indexing is if you have a Task that wraps a general purpose CLI (e.g. https://github.com/tektoncd/catalog/tree/main/task/gcloud/0.1) - if the CLI could be doing all kinds of different things (e.g. `gcloud projects list` vs `gcloud container clusters list`), then you might want to provide results that contain the complete stdout of the CLI (and in order to consume this in downstream tasks (with minimal effort), maybe you can index into it (the existing gcloud task provides no results at all)

for example:

```
# gcloud container clusters list
NAME        LOCATION     MASTER_VERSION   MASTER_IP     MACHINE_TYPE   NODE_VERSION         NUM_NODES  STATUS
eucalyptus  us-central1  1.18.20-gke.501  1.1.1.1  n1-standard-4  1.16.15-gke.6000 **  3          RUNNING
```

That would require a double index which this is not proposing but would be a logical next step (e.g. `$tasks.gcloud.results.[1][1]` if you wanted the location of the first cluster - _why the first??? that's where this isn't super compelling_) OR even to get the entire line and provide it as an array param - also not supported in this TEP but would require indexing support - `$tasks.gcloud.results[1]`)

bobcatfish commented:
One more use case came to mind for array indexing: if we add support for array results before we add support for looping (which as discussed is far from a sure thing and may never be added), there will be no way to use an array result from one Task with a Task that expects a string parameter.

For example, without looping, folks can get the reuse they want by hard coding the Tasks explicitly, which I'm guessing would probably support a surprising number of use cases (but look ugly and involve a lot of copy and paste), e.g.:

* Have a task that divides tests into 5 shards - if you know there are always 5 shards, you could add 5 go-test pipeline tasks and have each one use the result of the task that divdes them
* If you know a repo contains 14 Dockerfiles (e.g. [tekton/images in plumbing contains 14 dockerfiles](https://github.com/tektoncd/plumbing/tree/main/tekton/images) to build, you could have a task that grabs the 14 paths, and then feed them into 14 kaniko pipeline tasks individually

I'll add this as a use case as well; the TL;DR is that without looping support and without indexing, there will be no way to mix array results with tasks that expect strings

sbwsg commented:
Heh, that made me think of one more potential feature addition: a `.length` variable for the array param/result.

If we had that we could extend the support for a fixed number of tasks (your example of 5 test shards or 14 docker builds) and guard each one individually with a when expression to support dynamic array length:

```yaml
# guard kaniko-build-3 task - only runs if there are 3 or more entries in the array
when:
- input: $(params.dockerfiles.length)
  operator: In
  values:
  - "3"
  - "4"
  - "5"
  - #... 6, 7, 8, 9, 10, 11, 12 ...
  - "13"
  - "14"
# guard kaniko-build-4 task - only runs if there are 4 or more entries in the array
when:
- input: $(params.dockerfiles.length)
  operator: In
  values:
  - "4"
  - "5"
  - #... 6, 7, 8, 9, 10, 11, 12 ... 
  - "13"
  - "14"
```

This seems silly tho right?

bobcatfish commented:
> This seems silly tho right?

It doesn't seem totally silly - to me it reminds me of the [github actions matrix syntax](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix), specifically  [this article about running tests in parallel with github actions](https://rubyyagi.com/how-to-run-tests-in-parallel-in-github-actions/) where the best way to do something across 2 nodes seems to be to provide explicit indexes for each, e.g.:

```
      matrix:
        # Set N number of parallel jobs you want to run tests on.
        # Use higher number if you have slow tests to split them on more parallel jobs.
        # Remember to update ci_node_index below to 0..N-1
        ci_node_total: [2]
        # set N-1 indexes for parallel jobs
        # When you run 2 parallel jobs then first job will have index 0, the second job will have index 1 etc
        ci_node_index: [0, 1]
```

> Oh! I see Andrea already discussed determining array size in a comment on the doc.

Oh thanks for pointing that out, I had missed that in https://github.com/tektoncd/community/pull/477#discussion_r681006917

My first reaction was "ah yes, let's include the array length as variable replacement as well, but then I started wondering, wait, will we need something similar for dictionaries, and then i wondered about arrays of arrays, where you might want the size of the outer array, and the nested arrays... and now I'm thinking a) maybe jsonpath can help us here (but if not im leery of including our own expressions/functions around this, CEL could be another way to do this) and b) maybe we leave that out for now and revisit as we talk about adding more jsonpath, jsonschema (and CEL) support in the future?  (ill add this as an alternative at least)

sbwsg commented:
Agreed - leaving it out at this stage feels like the right move. 👍

bobcatfish commented:
Thanks for all the feedback so far @sbwsg @pierretasci @vdemeester @afrittoli !! I've added a bunch of updates, PTAL 🙏

sbwsg commented on teps/0076-array-result-types.md:
````diff
@@ -0,0 +1,531 @@
+   eucalyptus  us-central1  1.18.20-gke.501  1.1.1.1  n1-standard-4  1.16.15-gke.6000 **  3          RUNNING
+   ```
+
+   Using this as a result would require which this TEP is not proposing but would be a logical next step (e.g.
````
nit: missing word I think. `Using this as a result would require [something] which this TEP is not`

bobcatfish replied:
whoops thanks @sbwsg !! i think [something] is 'supporting nested arrays'

bobcatfish commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,563 @@
+* We've added json specific support to other parts of our APIs (e.g. [in triggers](https://github.com/tektoncd/triggers/blob/f2380d7d33666547ee7ed965730cf7dbc1aa5421/docs/triggerbindings.md#accessing-data-in-http-json-payloads))
+* Kubernetes supports `yaml` and `json`, so it makes sense to narrow our choices to at least those two, as opposed to
+  introducing a new format into the mix.
+
```
@mattmoor pointed out today that another good option would be to adopt [the open api spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md) that CRDs use to day

TODO: explore as an alternative

bobcatfish replied:
ah this actually makes more sense as a json schema alternative - going to add a "why json schema" section to https://github.com/tektoncd/community/pull/479 - and TL;DR it seems like Open API schemas might be a slightly better (more consistent with k8s and with what we want) approach for schemas but we don't need to cross that bridge until we expand our schema support

bobcatfish commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,563 @@
+  results. The proposal to add dictionary is mostly motivated by use cases for dictionary
+  results (vs params) - but today we only allow string results. It seems confusingly
+  inconsistent to imagine adding dictionary results but not array results (even though
+  we support array params)
```
TODO: rewrite this block its very confusing

bobcatfish commented:
ready for another look @sbwsg @pierretasci @vdemeester @jerop

bendory commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,544 @@
+    build, you could have a task that grabs the 14 paths, and then feed them into 14 kaniko pipeline tasks individually
+
+
+1. **Use case for array indexing: Tasks that wrap CLIs**
```
nit: this renders as `1.` when `3.` is expected, not sure why that's happening

pritidesai commented:
Overall, this TEP is proposing two features:

1. Allowing `params` to index into arrays
2. Allowing a task to write `JSON` blob into a result file which will then be converted into an array such that a consuming task can index into that result

No new types are introduced in general. Both the features will be limited to `string` type and array of `string`.

Both the ideas are great and I am almost onboard but trying to wrap my head around these and also understand them along with the other proposal, [TEP-0075](https://github.com/tektoncd/community/blob/8c09eccf29a829add9d3b853b84b73d32d557ad3/teps/0075-object-param-and-result-types.md#tep-0075-objectdictionary-param-and-result-types).

Feature (1) is no doubtably missed by many existing use cases when a `run` (pipelineRun/taskRun) specifies an array in a param, there is no way to refer to an individual array elements. The array param is only accessible via star notation `[*]` which returns the entire array. We could split this proposal to expedite the review/implementation process (just a thought).

For the second feature, the results are generally written as a `string` to the result file: 

```shell
echo -n "alpine-git-nonroot/Dockerfile, buildx-gcloud/Dockerfile, coverage/Dockerfile"  |  tee $(results.myResult.path)
```

How do we envision a `list of string` represented in a `JSON` format?

Also, in addition to the list of dockerfiles, there must be other information which needs to be communicated to the `build` task and in that case, just the list of dockerfiles wouldn't be sufficient. 

> result: list of starting and ending indexes for each shard, e.g. [[0, 7], [8, 14], [15, 23]]

Wouldn't this be considered as a nested arrays? 

When a task is allowed to produce a result in an array type, can the same result be consumed by pipeline result? If so, pipeline result will need an update as well. Also, we will have to support array indexing in pipeline result and implement a way to map an array element from a task result to a pipeline result. 

Are we envisioning adding support for other types, such as numeric, boolean? and similarly an array of those types?

afrittoli reviewed (commented):
Thanks for the updates @bobcatfish .
I'm not sure why this is marked on hold tough.
There will be a few details on the scope to be finalised in the implementation stage (indexing, lenght) but I feel like this waiting much longer than enough now.

/approve

pritidesai commented:
hey @bobcatfish I was hoping to hear your response for the [comments](https://github.com/tektoncd/community/pull/477#issuecomment-922065236) ⬆️  I had left before we can merge this. At the same time, we have approvals from most of the reviewers so please feel free to merge it.

mogsie commented:
Have you considered [JSON Text Sequences](https://datatracker.ietf.org/doc/html/rfc7464) (simply put, concatinated JSON documents) as the format of an array?  These text sequences have a nice property of being able to concatinate, so instead of `[ "foo", "bar", "baz" ]` — a JSON array, you would have
```
"foo"
"bar"
"baz"
```
this can be composed by e.g. concatinating results in the pipeline:

```
params:
  - name: an-array
    value: |
      $(tasks.foo.results.thing)
      $(tasks.bar.results.thing)
```

If these results are both `[ ... ]` syntax, then this composability wouldn't work.

bobcatfish commented:
Hey @mogsie ! Thanks for the suggestion - Looking at the RFC, am I right that this will _only_ work for arrays, and not for more complex json types (e.g. objects - which I'm proposing support for in https://github.com/tektoncd/community/pull/479) and later nested types (e.g. arrays of arrays)? If so then I think we (at least initially) wouldn't want this to be our format for arrays b/c we'd want to choose a format that will work for these other types as well (but we could potentially adds support for something like this as an optimization later?)

But let me know if I'm totally wrong and this will work for more complex types as well!

bobcatfish commented:
@pritidesai  finally addressing your points in https://github.com/tektoncd/community/pull/477#issuecomment-922065236 :

> We could split this proposal to expedite the review/implementation process (just a thought).

That's a great point @pritidesai  XD If I had done that earlier then maybe we'd have that TEP accepted already (or maybe not since it took me so long to get back to these X'D) - i can still do this if needed, but I get the impression that folks are nearly ready to accept this TEP as-is so if we can keep it bundled together I'd prefer that. (I think to me this TEP was kind of about 'missing features in array param support' in general) I also think indexing is somewhat related to array result support in that without it, array results will only work with tasks that use array params (and never with string params)

> How do we envision a list of string represented in a JSON format?

This would look like a JSON array, for example:

```
["cat", "dog", "squirrel"]
```

Also, in addition to the list of dockerfiles, there must be other information which needs to be communicated to the build task and in that case, just the list of dockerfiles wouldn't be sufficient.

> Also, in addition to the list of dockerfiles, there must be other information which needs to be communicated to the build task and in that case, just the list of dockerfiles wouldn't be sufficient.

very true! I think [TEP-0075](https://github.com/tektoncd/community/blob/8c09eccf29a829add9d3b853b84b73d32d557ad3/teps/0075-object-param-and-result-types.md#tep-0075-objectdictionary-param-and-result-types) - and eventually adding support for nested types will fill this need nicely - for example, a result that is an array of objects where each object contains the name of the dockerfile and whatever other info you need

I do think that just array support will be enough to support a significant number of use cases like this though (with some finessing) - but like you said it's not the whole picture

> result: list of starting and ending indexes for each shard, e.g. [[0, 7], [8, 14], [15, 23]]
> Wouldn't this be considered as a nested arrays?

yes and I think that supporting that is definitely the inevitable next step here - we could include support for nested arrays right off the bat but i think our path forward is much simpler if we tackle the nested types question as a whole post TEP-0075 - but I'll update this example to try to make that clear and show a way to do it initially without nesting

> When a task is allowed to produce a result in an array type, can the same result be consumed by pipeline result? If so, pipeline result will need an update as well

yes, I'll add that as an explicit requirement

> Are we envisioning adding support for other types, such as numeric, boolean? and similarly an array of those types?

maybe! i could include something explicit about that but I'm not sure that I want to jump to that conclusion right away, I think we should consider each type on a case by case basis

bobcatfish commented:
I've finally given this a long overdue update and I think I've addressed all the issues raised so far - PTAL @vdemeester @afrittoli @pierretasci @jerop @sbwsg !

I think @sbwsg @vdemeester and @afrittoli have already approved so looking for approval from @pritidesai @pierretasci and @jerop (its been a long time so just let me know if you don't want to be assigned any longer) - and I'll be trying to follow this up quickly with a PR to propose moving to implementable.

mogsie commented:
@bobcatfish, I think it would work fine with both arrays of strings, arrays of objects, arrays of arrays, essentially everything that can be expressed in s JSON Array would be possible to be expressed in a document of type _json text sequence_

Example json array (the current proposal, after _dictionaries_ have been addedI believe):
```
[ 1, 2, 3, "four", { "five": 5 }, [6], { "seven": [ 7, 7, "7" ] } ]
```

Identical data, but expressed as a JSON text sequence:
```
1
2
3
"four"
{ "five": 5 }
[6]
{ "seven": [ 7, 7, "7" ] }
```
The newlines here are significant.

Using a _json text sequence_ is not in any way incompatible with any future "object" notation, IMHO it would fit like a glove.  You could define a result as a dictionary with two keys, e.g. _repo_ and _sha_:

```
{"repo":"foo", "sha": "abcde" }
```

If this field were defined to be an _array_ then the above would essentially be _an array with one element_.  If the task returned an _array_ of this dictionary, then it might express it like this (three items in the "array"):

```
{"repo":"foo", "sha": "abcde" }
{"repo":"the_second_repo", "sha": "bf3ab" }
{"repo":"third", "sha": "87cd2" }
```

The json text sequence format has a number of nice properties:

1. It is compatible with the proposed object syntax (as shown above) — but does not require it (the proposal could require only primitive json types
2. it has a way of expressing strongly typed values like string, boolean, and int
3. it is composable — I can take two results and concatinate them together to get a longer "array" containing all the elements.
4. empty values represent exactly that.
5. No "special syntax" is needed to distinguish between (a) a list of items, and (b) a single item.  Indeed, you can take two "single value" results, concatinate them, and you get a list consisting of two items automatically.

This would mean that the main data types should be expressed as JSON primitives, (i.e. that if `xyzzy` is defined as a result of a task, then `echo "foo" > $(results.xyzzy)` would result in the following internal representation:

```
"foo\n"
```

I believe that storing results in terms of _json primitives_ is the essence here; _json text sequences_ of these are then the natrual "array" syntax, and _complex objects_ and schemas can be added on top of that.

bobcatfish commented:
Ahh I see, thanks for explaining in so much detail @mogsie . How strongly do you feel about the json text sequence syntax being supported right from the start? If I'm understanding correctly it seems like a) we'll likely want support for plain old json regardless of whether we support text sequences or not (we need to parse the content of each line anyway which is itself a complete json type) and b) this can be additively supported in a backwards compatible way.

My preference is to keep this as simple as we can in the initial iterations (which is why I don't even want to support type nesting initially!) and expand the functionality later, including potentially adding support for json text sequences. What do you think?

I've added JSON text sequences as an alternative in the proposal in the meantime - and I've tried to capture the properties you listed above

> empty values represent exactly that.

I don't quite follow that one - is this an easier way to specify 'null'?

pierretasci commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,620 @@
+* Kubernetes supports `yaml` and `json`, so it makes sense to narrow our choices to at least these, as opposed to
+  introducing a new format into the mix.
+
+We may want to consider also adding support for [JSON text sequences](#also-support-json-text-sequences) in future.
```
Other important note, YAML is a superset of json, all valid json is valid yaml so in a sense, we get the best of both 😄

bobcatfish replied:
Good point! I'll add that as a bullet as well :D

pierretasci commented:
/approve

mogsie commented:
I understand that supporting object syntax is over-the-top for this "only arrays" TEP.  I initially suggested just that, a pure "array of strings" syntax, but the point about "future proofness" of this didn't quite come across.  My main reasoning was that if you serialize an array of strings using JSON Text Sequence, then you gain a lot of future compatibility with TEP-0075, you get to be able to join two arrays by concatinating them, and so on.

So instead of a result array being serialized here depicted as a "value" in yaml as a string representation of a JSON array

```
value: |
  [ "one", "two", "three" ]
```

I suggest using this (json text sequence compatible):

```
value: |
  "one"
  "two"
  "three"
```

> How strongly do you feel about the json text sequence syntax being supported right from the start? 

I'm just concerned that Tekton might miss out on opportunities if it selects an array syntax that isn't easily composable / needs yet another syntax for object notation.  e.g. if you have a set of tasks, and each one emits a _result_ and then you want to join these results together to form an array; if results are emitted as "single element json text sequences", they can just be joined together:

```
task 1 result = "foo"\n
task 2 result = "bar"\n
task 3 result = "baz"\n

input to some array = task1 result + task2 result + task 3 result =
"foo"
"bar"
"baz"

-- an array with three elements in it
```

> empty values represent exactly that.

I think I was talking about how when value is empty, (i.e. an _empty array_) it is just an empty string.

> I don't quite follow that one - is this an easier way to specify 'null'?

Null is just the word "null" without quotes, same as true, false:  Here the second value is a null element

```
value: |
  "one"
  null
  "three"
```

mogsie commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,620 @@
+[TEP-0075](https://github.com/tektoncd/community/pull/479), dictionaries (technically "objects"). Possible other formats
+we could use are:
+
+* `yaml` (i dont think there is a syntax like jsonpath available for yaml? i.e. it'd be harder to expose via variable
```
yaml ends up being in-memory, a memory structure that's very similar to json, so you should be able to use jsonpath / jsonpointer etc on a yaml document.

mogsie commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,620 @@
+    steps: ...
+    results:
+    - name: animals
+      type: array # this field is new for results
```
I know the whole proposal is about making array results, but this "type: array" is in conflict with "type: object" — have you considered a different attribute like "multiple: true" or something?  

With both 0075 and 0076, you'd be able to define something like:

```
  name: animals
  type: object
  properties:
    name:
      type: string
    weight:
      type: number
  multiple: true
```

and emit (using json text sequence, of course)

```
cat <<EOF > $(results.animals.path)
{ "name": "cat", weight: 33}
{ "name": "dog", weight: 44}
```

bobcatfish replied:
Interesting - i was imagining the syntax being more json schema aligned to support the use case you're describing (an array of objects), e.g. something like this (based on https://json-schema.org/understanding-json-schema/reference/array.html and https://json-schema.org/learn/miscellaneous-examples.html#arrays-of-things):

```
  name: animals
  type: array
  items: {
    type: object
    properties:
      name:
        type: string
      weight:
        type: number
  }
```

And if we supported json text sequences, i would think the schema would look the same, the difference would be in how the results written by the Task were formatted? (i.e. im not seeing these features as being in conflict but let me know if I'm missing something)

However I have intentionally tried to keep nested types out of the scope of this proposal and TEP-0075 as well. I would want to make sure nothing we propose here would stop us from adding nested type support in the future so if you see this being incompatible please continue to raise it.

Personally I would still lean toward using the json schema syntax for this vs. introducing a new `multiple` field.

bobcatfish commented:
> I initially suggested just that, a pure "array of strings" syntax, but the point about "future proofness" of this didn't quite come across. My main reasoning was that if you serialize an array of strings using JSON Text Sequence, then you gain a lot of future compatibility with TEP-0075, you get to be able to join two arrays by concatinating them, and so on.

Thanks for clarifying @mogsie  - I've tried to add more detail to the section I added on JSON text sequences to try to capture how this syntax could make composability easier for Tasks in a Pipeline, let me know if I'm still not capturing it (and feel free if you want to write something verbatim to add into the proposal, I'd be happy to add you as an author if you want).

My preference still remains to keep this out of scope from the initial proposal - provided none of the choices we make here preclude adding support in the future.

jerop commented on teps/0076-array-result-types.md:
```diff
@@ -0,0 +1,637 @@
+   strings.
+
+   If we add support for array results before we add support for looping (and until we have an approved proposal, there is
+   no guarantee we ever will!) there will be no way to use an array result from one Task with a Task that expects a
```
getting there soon 😉 

https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md

dibyom commented:
/cc @pritidesai to review

pritidesai commented:
thank you @bobcatfish for addressing my comments! 
/lgtm

@mogsie can this be merged given that your proposal is added as a possible extension to the existing proposal?

pritidesai commented:
/hold cancel

pritidesai commented:
API WG - ready to merge, cancelling the hold

xchapter7x commented:
/area s3c
